windows-gaming-driver/src/setup/usb.rs

```diff
@@ -82,18 +84,116 @@ pub fn select(machine: &mut MachineConfig) -> bool {
+    let devs = list_devices(None).expect("Can't read connected USB devices");
+
+    loop {
+        println!("You have currently selected the following usb devices: ");
+        let mut last = 0;
+        for (i, binding) in usb_devices.iter().cloned().enumerate() {
+            if let Some(pos) = devs.iter().position(|dev| UsbBinding::from(dev.id) == binding
+                    || UsbBinding::from(dev.port) == binding) {
+                // device is connected
+                let dev = &devs[pos];
+                let name = to_name(dev.vendor_name.as_ref(), dev.product_name.as_ref());
+                println!("[{}]\t{} [{:04x}:{:04x}]", i, name, dev.id.vendor, dev.id.product);
+            } else {
+                // device not connected, handle cases of by-id and by-port
+                match binding {
+                    UsbBinding::ById(DeviceId { vendor, product }) => {
+                        let (vendor_name, product_name) = hwid::hwid_resolve_usb(vendor, product).ok().unwrap_or((None, None));
+                        let name = to_name(vendor_name, product_name);
+                        // TODO: Find out HidKind from vendor and product
+                        println!("[{}]\t{} [{:04x}:{:04x}] (Not Connected)", i, name, vendor, product);
+                    },
+                    UsbBinding::ByPort(DevicePort { bus, port }) => {
+                        println!("[{}]\tBus {} Port {} (Not Connected)", i, bus, port);
+                    }
+                }
+            }
+            last = i;
+        }
+        println!("[{}]\tNone", last + 1);
+        let index = ask::numeric("Which device would you like to remove?", 0..(last + 2));
+        if index == last + 1 {
+            break;
+        }
+        usb_devices.swap_remove(index);
+    }
+    let mut mouse = None;
+    let mut mouse_idx = None;
+    let mut keyboard = None;
+    for (i, binding) in usb_devices.iter().cloned().enumerate() {
+        if let Some(idx) = devs.iter().position(|dev| UsbBinding::from(dev.id) == binding
+                || UsbBinding::from(dev.port) ==  binding) {
+            let dev = &devs[idx];
+            if dev.kind == Some(HidKind::Mouse) && mouse == None {
+                mouse = Some(binding);
+                mouse_idx = Some(i);
+            }
+            if dev.kind == Some(HidKind::Keyboard) && keyboard == None {
+                keyboard = Some(binding);
+            }
+        }
+    }
+    if let Some(ix) = mouse_idx {
+        usb_devices.swap(0, ix);
+    }
+    (mouse, keyboard)
+}
+
+fn to_name<T: AsRef<str>>(vendor_name: Option<T>, product_name: Option<T>) -> String {
+    match (vendor_name, product_name) {
+        (None, None) => "Unknown vendor Unknown product".to_string(),
+        (Some(vendor), None) => format!("{} Unknown product", vendor.as_ref()),
+        (Some(vendor), Some(product)) => format!("{} {}", vendor.as_ref(), product.as_ref()),
+        (None, Some(_)) => unreachable!()
```

Comment:
If you're matching like this *anyways*, why not just keep the `Option<(String, Option<String>)`? Much less awkward.